File: src/CSharp/SodaFlow.Tests.Internal/CellTests.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using SodaFlow.Functional;
using TUnit.Assertions;
using TUnit.Assertions.Enums;
using TUnit.Assertions.Extensions;
using TUnit.Core;

namespace SodaFlow.Tests.Internal;

public sealed class CellTests
{
    [Test]
    public async Task TestTransaction()
    {
        bool calledBack = false;

        TransactionInternal.Apply((trans, _) =>
        {
            trans.Prioritized(node: Node<Unit>.Null, action: _ => calledBack = true);
            return UnitInternal.Value;
        });

        await Assert.That(calledBack).IsTrue();
    }

    [Test]
    public async Task TestRegen()
    {
        List<int> @out = [];

        TransactionInternal.Apply((trans, _) =>
        {
            SetNeedsRegeneratingAndPrioritized(() => @out.Add(1));
            SetNeedsRegeneratingAndPrioritized(() => SetNeedsRegeneratingAndPrioritized(() => @out.Add(4)));
            SetNeedsRegeneratingAndPrioritized(() => @out.Add(2));

            SetNeedsRegeneratingAndPrioritized(() =>
                SetNeedsRegeneratingAndPrioritized(() => SetNeedsRegeneratingAndPrioritized(() => @out.Add(6))));

            SetNeedsRegeneratingAndPrioritized(() => SetNeedsRegeneratingAndPrioritized(() => @out.Add(5)));
            trans.Prioritized(node: new Node<Unit>(), action: _ => @out.Add(3));

            return UnitInternal.Value;

            void SetNeedsRegeneratingAndPrioritized(Action action) =>
                trans.Prioritized(node: new Node<Unit>(), action: _ => action());
        });

        await Assert.That(@out).IsEquivalentTo([1, 2, 3, 4, 5, 6], CollectionOrdering.Matching);
    }
}

File: src/FSharp/SodaFlow.FSharp.Tests.Internal/CellTests.fs
module SodaFlow.Tests.Internal.Cell

open System.Collections.Generic
open SodaFlow
open SodaFlow.Tests
open TUnit.Core

type ``Cell Tests``() =

    [<Test>]
    member _.``Test Transaction``() =
        task {
            let mutable calledBack = false
            TransactionInternal.Apply(fun trans _ -> trans.Prioritized(Node<unit>.Null, fun trans -> calledBack <- true))
            do! Expect.True calledBack
        }

    [<Test>]
    member _.``Test Regen``() =
        task {
            let out = List<_>()

            TransactionInternal.Apply(fun trans _ ->
                let setNeedsRegeneratingAndPrioritized a =
                    trans.Prioritized(Node<unit>(), (fun _ -> a ()))

                setNeedsRegeneratingAndPrioritized (fun () -> out.Add 1)
                setNeedsRegeneratingAndPrioritized (fun () -> setNeedsRegeneratingAndPrioritized (fun () -> out.Add 4))
                setNeedsRegeneratingAndPrioritized (fun () -> out.Add 2)

                setNeedsRegeneratingAndPrioritized (fun () ->
                    setNeedsRegeneratingAndPrioritized (fun () ->
                        setNeedsRegeneratingAndPrioritized (fun () -> out.Add 6)))

                setNeedsRegeneratingAndPrioritized (fun () -> setNeedsRegeneratingAndPrioritized (fun () -> out.Add 5))
                trans.Prioritized(Node<unit>(), fun _ -> out.Add 3))

            do! Expect.Sequence([ 1; 2; 3; 4; 5; 6 ], out)
        }
